Added in Meteor.loggingOut() and related Blaze helpers.

hwillson opened the pull request:
Hi all - this PR is intended to help address issues #769 and #1331. It's essentially a re-working of the very old PR #826 (submitted by @raix - thanks!). I've modified it to work with the current codebase and added some tests. Let me know if you have any questions / need anything changed. Thanks!

hwillson commented:
I forgot to mention - I created a [small runnable app](https://github.com/hwillson/meteor-issue-769) that shows the `loggingOut` helper firing (along with `Meteor.loggingOut()`), in-case anyone wants to see these changes in action. Thanks!

raix commented:
Thanks, 4 years old :)

hwillson commented:
Hehe, I know - we're in the process of cleaning up old issues @raix so this was uncovered. There is an amazing amount of good stuff, like your changes, that got overlooked! Hopefully more of it starts to filter in soon. Thanks again!

raix commented on packages/accounts-base/accounts_client.js:
```diff
@@ -17,6 +17,9 @@ export class AccountsClient extends AccountsCommon {
 
+    this._loggingOut = false;
+    this._loggingOutDeps = new Tracker.Dependency;
+
```
maybe use `ReactiveVar`:
```js
this._loggingOut = new ReactiveVar(false)
```

hwillson replied:
Thanks for the feedback @raix! I didn't use a `ReactiveVar` since these changes were modelled after the existing `loggingIn` code. That being said, using a `ReactiveVar` would definitely clean the code up a bit. I'll look into making these changes, and consider updating the code for `loggingIn` as well (and add some new tests for that part of the code since there don't seem to be any). Thanks again for taking a look at this!

raix replied:
Good point

raix commented on packages/accounts-base/accounts_client.js:
```diff
@@ -49,6 +52,13 @@ export class AccountsClient extends AccountsCommon {
+    if (this._loggingOut !== x) {
+      this._loggingOut = x;
+      this._loggingOutDeps.changed();
+    }
```
`this.loggingOut(x);`

raix replied:
Actually just skip this helper function

raix commented on packages/accounts-base/accounts_client.js:
```diff
@@ -59,15 +69,26 @@ export class AccountsClient extends AccountsCommon {
+   */
+  loggingOut() {
+    this._loggingOutDeps.depend();
+    return this._loggingOut;
```
`return this._loggingOut.get();`

raix commented on packages/accounts-base/accounts_client.js:
```diff
@@ -59,15 +69,26 @@ export class AccountsClient extends AccountsCommon {
    */
   logout(callback) {
     var self = this;
+    self._setLoggingOut(true);
```
`self._loggingOut.set(true);`

raix commented on packages/accounts-base/accounts_client.js:
```diff
@@ -59,15 +69,26 @@ export class AccountsClient extends AccountsCommon {
     self.connection.apply('logout', [], {
       wait: true
     }, function (error, result) {
+      self._setLoggingOut(false);
```
and we could add `.set` here too

hwillson commented:
I updated things a bit to leverage `ReactiveVar`'s for both `loggingIn` and `loggingOut`, and added some additional tests for the `loggingIn` side of things. Thanks all!

abernix commented:
I'll try to review this a bit more in detail during/before next week's issue club (couldn't get to it yesterday) but first pass doesn't.  My only initial thought might be any possibility of a term aside from `loggingInOut` for the helper?  Possibly something along the lines of `authChanging`, but maybe even just adding the `Or` in there (`loggingInOrOut`)?

hwillson commented:
Sounds good @abernix - it's been 4 years, so no rush! 😄 I like your `loggingInOrOut` suggestion more than `authChanging`; `authChanging` might be a bit too ambiguous, since auth can be taken to mean different things (like "authentication" or "authorization"). I'll adjust - thanks again!

benjamn commented on packages/accounts-base/accounts_client.js:
```diff
@@ -438,4 +453,24 @@ if (Package.blaze) {
+   * @isHelper true
+   * @summary Calls [Meteor.loggingOut()](#meteor_loggingout).
+   */
+  Package.blaze.Blaze.Template.registerHelper('loggingOut', function () {
```
I realize there is already code in this file that assumes `Package.blaze` is defined, but I think I would prefer if we were a bit more defensive about checking whether `blaze` is installed, since it's only a weak dependency of `accounts-base`.

hwillson replied:
Sounds good - I'll get this resolved - thanks!

hwillson replied:
Just reviewing this now - it looks like there is a [`Package.blaze` check just above the helper registration code](https://github.com/hwillson/meteor/blob/9b8f5a87864e2e6399964957bbeef32b8ca5606f/packages/accounts-base/accounts_client.js#L436). Do you think this is defensive enough? I can expand this further by converting the `Package.blaze` check into a `Package.hasOwnProperty('blaze')` check, paired with some additional checks to make sure the blaze object contains a `Blaze` property, but this might be overkill (I think it's pretty safe to assume `Package.blaze` really only points to a valid blaze object, if truthy). Let me know though - thanks!

mitar replied:
I think `Package.blaze` is good enough. This is what everyone is doing for weak dependencies.
